annotated.go

```diff
@@ -89,3 +90,228 @@ func (a Annotated) String() string {
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If there are more Annotations are specified than than one per
+// Annotation type, none of the Annotations will be applied.
+//
+// For example,
+//
+// fx.Provide(
+//   fx.Annotate(
+//     NewGateWay,
+//     fx.ParamTags(`name:"ro" optional:"true"`),
+//     fx.ParamTags(`name:"rw"),
+//     fx.ResultTags(`name:"foo"`)
+//   )
+// )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+
+	if !verifyAnnotation(numIn, numOut, anns...) {
+		return f
+	}
+
+	var ins []reflect.Type
+	var outs []reflect.Type
+
+	annotatedIn := false
+	annotatedOut := false
+
+	for _, ann := range anns {
+		if pTags, ok := ann.(paramTags); ok {
+			ins = pTags.getAnnotatedType(fType)
+			annotatedIn = true
+		}
+		if rTags, ok := ann.(resultTags); ok {
+			outs = rTags.getAnnotatedType(fType)
+			annotatedOut = true
+		}
+	}
+
+	if !annotatedIn {
+		ins = make([]reflect.Type, numIn)
+		for i := 0; i < fType.NumIn(); i++ {
+			ins[i] = fType.In(i)
+		}
+	}
+	if !annotatedOut {
+		outs = make([]reflect.Type, numOut)
+		for i := 0; i < fType.NumOut(); i++ {
+			outs[i] = fType.Out(i)
+		}
+	}
+
+	var newF func([]reflect.Value) []reflect.Value
+
+	if annotatedIn && annotatedOut {
+		newF = func(args []reflect.Value) []reflect.Value {
+			fParams := make([]reflect.Value, numIn)
+			params := args[0]
+			for i := 0; i < numIn; i++ {
+				fParams[i] = params.Field(i + 1)
+			}
+			fResults := fVal.Call(fParams)
+			results := reflect.New(outs[0]).Elem()
+			for i := 0; i < numOut; i++ {
+				results.FieldByName(fmt.Sprintf("Field%d", i)).Set(fResults[i])
+			}
+			return []reflect.Value{results}
+		}
+	} else if annotatedIn {
+		newF = func(args []reflect.Value) []reflect.Value {
+			fParams := make([]reflect.Value, numIn)
+			params := args[0]
+			for i := 0; i < numIn; i++ {
+				fParams[i] = params.Field(i + 1)
+			}
+			return fVal.Call(fParams)
+		}
+	} else { // annotatedOut
+		newF = func(args []reflect.Value) []reflect.Value {
+			fResults := fVal.Call(args)
+			// wrap the result in annotated struct
+			results := reflect.New(outs[0]).Elem()
+			for i := 0; i < numOut; i++ {
+				results.FieldByName(fmt.Sprintf("Field%d", i)).Set(fResults[i])
+			}
+			return []reflect.Value{results}
+		}
+	}
+
+	annotatedFuncType := reflect.FuncOf(ins, outs, false)
+	annotatedFunc := reflect.MakeFunc(annotatedFuncType, newF)
+	return annotatedFunc.Interface()
+}
+
+func verifyAnnotation(numIn int, numOut int, anns ...Annotation) bool {
+	sawParamAnn := false
+	sawResultAnn := false
```

Comment:
This seems like interpreting the annotations twice. There's an optional, slightly different way to go about this:

Instead of having paramTags and resultTags generate `[]reflect.Type`, have them update an annotations struct (like regular functional options), failing if the struct is already filled:

```
type annotations struct {
  ParamTags []string
  ResultTags []string
}

type Annotation interface{ apply(*annotations) }

func (pt paramTags) apply(ann *annotations) error {
  if len(pt.ParamTags) > 0 { return fmt.Errorf(...) }
  // ...
}
```

Then in `func Annotate`, as we apply these options, we'll get back an error with some amount of information.

We can generate the `[]reflect.Type`s inside `func Annotate` as needed (or inside a helper) elsewhere, probably with some shared code.